test/lit/passes/code-pushing_into_if.wast

```diff
@@ -0,0 +1,540 @@
+      (drop (local.get $x))
+    )
+  )
+
+  ;; CHECK:      (func $past-condition-no (param $p i32)
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (local $t i32)
+  ;; CHECK-NEXT:  (local.set $x
+  ;; CHECK-NEXT:   (local.get $t)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (local.tee $t
+  ;; CHECK-NEXT:    (local.get $p)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $past-condition-no (param $p i32)
+    (local $x i32)
+    (local $t i32)
+    ;; We cannot push this due to the tee in the if condition.
+    (local.set $x (local.get $t))
+    (if
+      (local.tee $t (local.get $p))
+      (drop (local.get $x))
+    )
+  )
+
+  ;; CHECK:      (func $past-condition-no-2
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (local $t i32)
+  ;; CHECK-NEXT:  (local.set $x
+  ;; CHECK-NEXT:   (local.get $t)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (local.get $x)
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $past-condition-no-2
+    (local $x i32)
+    (local $t i32)
+    ;; We cannot push this due to the read of $x in the if condition.
+    (local.set $x (local.get $t))
+    (if
+      (local.get $x)
+      (drop (local.get $x))
+    )
+  )
+
+  ;; CHECK:      (func $past-condition-no-3 (param $p i32)
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (local $t i32)
+  ;; CHECK-NEXT:  (local.set $x
+  ;; CHECK-NEXT:   (local.get $t)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (local.tee $x
+  ;; CHECK-NEXT:    (local.get $p)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $past-condition-no-3 (param $p i32)
+    (local $x i32)
+    (local $t i32)
+    ;; We cannot push this due to the write of $x in the if condition.
+    (local.set $x (local.get $t))
+    (if
+      (local.tee $x (local.get $p))
+      (drop (local.get $x))
+    )
+  )
+
+  ;; CHECK:      (func $if-condition-return (param $p i32)
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (block (result i32)
+  ;; CHECK-NEXT:    (return)
+  ;; CHECK-NEXT:    (local.get $p)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (block
+  ;; CHECK-NEXT:    (local.set $x
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $if-condition-return (param $p i32)
+    (local $x i32)
+    (local.set $x (i32.const 1))
+    (if
+      (block (result i32)
+        (return) ;; This return does not prevent us from optimizing; if it
+                 ;; happens then we don't need the local.set to execute
+                 ;; anyhow.
```

Comment:
What if this is not a return but a branch to an outer block after which the local is used?

Reply:
That's handled by the shared logic in `isPushable`, which rules out such cases. But good to test it here as well, I added a test now.